DbgProvider/public/Debugger/DbgEngDebugger.cs

```diff
@@ -61,6 +61,28 @@ internal static DbgEngDebugger NewDebugger()
                 } );
         }
 
+        public static DbgEngDebugger OpenDumpFile( string dumpFileName, string targetFriendlyName = "" )
+        {
+            return OpenDumpFileAsync( dumpFileName, targetFriendlyName ).Result;
```

Comment:
>.Result [](start = 72, length = 7)

Does this work as expected? Briefly looking over how `Mount-DbgDumpFile` works, it calls `OpenDumpFileWide` (on the dbgeng thread) and then calls into `base.ProcessRecord`, which pumps a message loop while waiting for the debug event. If we block the cmdlet thread on `.Result`, it won't be able to pump messages, and perhaps deadlock.

Oh wait... I see the difference: you call `WaitForEvent`, whereas `Mount-DbgDumpFile` calls `WaitForEventAsync`. Will debug events queue WriteObject calls to this thread? Or maybe they just go into a null pipe or something?

Reply:
In the context I have been using it with, it has been consistently working just fine for me - but there is no cmdlet thread nor any pipes to receive output, I definitely don't have a strong enough understanding of what the calls are actually doing to be confident about how it would behave in those cases.

Reply:
I went back to see what happens to notifications raised while waiting for a debug event (RealDebugEventCallbacks.cs) Can you check debugger.`HoldingPipelineCount` to see if you are accumulating stuff in there?

I've always had the idea that DbgShell could also be used as C# library code (as opposed to purely an interactive or scripted PowerShell thing), but never really tried to exercise it that way. So I'm glad that you're trying it out, and I'm also worried that the API is terrible for that sort of thing (sorry!). I think it would take some focused design work to get it into shape.

I'm trying to think of what I would want some XML doc comments on these methods to say, to explain how to use them correctly... but the thing is, I don't know if you actually CAN--for instance you can't set the current pipeline (`_SetCurrentPipeline` is internal), so I believe that anything written to the pipeline is doomed to hang out in the "holding" pipe forever. Maybe that's not a dealbreaker for your particular scenario, but it's definitely not good in general.

Because I don't think anybody could be using these methods well, I'm happy to take this change, but if you are serious about using dbgprovider.dll as a library, we should come up with a plan / design / strategy for how the API *should* work. I *think* I have the fundamentals in place (such as an abstraction of the PowerShell pipeline), but it's just not all exposed, and there has never been a real-life scenario (for me) driving it. If you were able to share more about your scenario, that would be helpful.


---
In reply to: [238537804](https://github.com/Microsoft/DbgShell/pull/49#discussion_r238537804) [](ancestors = 238537804)

Reply:
Looks like I'm getting one holding pipe entry per dump file I load (a message such as "(40fc.4568): Break instruction exception - code 80000003 (!!! second chance !!!)win32u!NtUserWaitMessage+0xc")

`debugger.HandleDbgEngOutput(line => Console.WriteLine(line));` got me all of the output I needed (mostly troubleshooting symbol load).

The API has generally covered everything I've needed at least as well as DbgEng (though perhaps not always well polished - UDTs in particular are a bit of a pain to work with, but so little of what I've been doing maps to useful type definitions that it hardly matters). I should be able to share more detail about what I'm doing privately; I'll see about how much & send you an e-mail.